lib/truffle/truffle/https_downloader.rb

```diff
@@ -0,0 +1,33 @@
+require 'net/http'
+
+module Truffle
+  module HTTPSDownloader
+    HTTP_PROXY_HEADER = "HTTP/1.0 200 Connection established\r\n\r\n"
+
+    class CurlResponse < Net::HTTPOK
+      def body
+        @body
+      end
+    end
+
+    def self.download(uri)
+      resp = CurlResponse.new('1.1', 200, 'OK')
+      raise 'curl is missing' unless (`curl --help` rescue nil)
+      resp_raw = `curl -i -s #{uri}`
+      index_offset = resp_raw.start_with?(HTTP_PROXY_HEADER) ? HTTP_PROXY_HEADER.size : 0
+      blank_line_idx = resp_raw.index("\r\n\r\n", index_offset)
+      header = resp_raw[0...blank_line_idx]
+      resp.body = resp_raw[(blank_line_idx+4)..-1]
+
+      if m = /ETag: (\"[[:alnum:]]*\")/.match(header)
+        resp['ETag'] = m[1]
+      end
+      resp
+    end
+
+    def self.download_to(uri, path)
+      raise 'curl is missing' unless (`curl --help` rescue nil)
```

Comment:
A couple of thoughts on this: 

- could we extract this out into a `DependencyException` or something along those lines? 

- I'd recommend using `command -v curl` to check for `curl`'s existence. There are quite a few performance related benefits associated with `command` over alternative methods, and I'd say that's something that matters, especially since we're already shelling out, you know?

- This is outside of the scope of this PR, but it may be a good idea to include a dependency check higher up the stack (for all dependencies), for clarity, as well as allowing shorter short-circuiting :)